### ISSUE TITLE
Images returned by tf.image.decode_jpeg have significant differences with images returned by other common loaders.

### ISSUE THREAD

**Issue by pEtienn** [tensorflow/tensorflow#102996]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

No

### Source

binary

### TensorFlow version

2.20

### Custom code

Yes

### OS platform and distribution

windows

### Python version

3.12.12

### Current behavior?

Images returned by io.read_file and tf.image.decode_jpeg are substantially different to images read by the following functions: 
>PIL.Image.open
>plt.imread 
>keras.utils.load_img

Those three functions return the exact same array while loading the JPG files I tried, while the function path_to_image  (using io.read_file and tf.image.decode_jpeg/image) returns substantially different images. While I don't notice the difference between images with my eyes, subtracting the images together makes the difference obvious. 

<img width="412" height="418" alt="Image" src="https://github.com/user-attachments/assets/a55953d0-e087-44fe-85e2-936b6f993f73" />

<img width="412" height="418" alt="Image" src="https://github.com/user-attachments/assets/0ae40e6e-5d5a-4ebf-9a6e-5d1d665da9c7" />

I noticed the issue while using the function tf.keras.utils.image_dataset_from_directory, which uses a function similar to the function path_to_image in the code snippet. I was getting unexpected results when predicting labels for a single image loaded with PIL, while my dataset was loaded with tf.keras.utils.image_dataset_from_directory. 

Another factor in the difference between tf.keras.utils.image_dataset_from_directory images and images from PIL is that tf.keras.utils.image_dataset_from_directory always resizes, even when there is no need. This can change a bit the image loaded, but in the code example I provided, there is no resizing. I don't think resizing is the main culprit behind the large differences I see.

The resizing issue was mentioned here: https://github.com/keras-team/keras/issues/16227
The difference between tf.keras.utils.image_dataset_from_directory images and PIL images were mentioned here: https://github.com/keras-team/keras/issues/16227https://discuss.ai.google.dev/t/why-does-image-dataset-from-directory-return-a-different-array-than-loading-images-normally/31378

### Standalone code to reproduce the issue

```shell
import tensorflow as tf
import numpy as np
import PIL
import matplotlib.pyplot as plt
import keras

def path_to_image(path, num_channels):
  img = io.read_file(path)
  img=tf.image.decode_jpeg(img,channels=num_channels)
  return img

imagePath=r"test_image.jpg"
imPIL=np.array(PIL.Image.open(imagePath),dtype=np.uint8)
imPLT=plt.imread(imagePath)
imTF=np.uint8(keras.utils.load_img(imagePath))
imTF2=path_to_image(imagePath,3)
print('The first 3 loaders load the exact same image: ',np.all((imPIL==imPLT)==(imPLT==imTF)))
print('The image loaded with decode_jpeg is equal to the others: ',np.all(imTF2==imTF))
```

### Relevant log output

```shell
The first 3 loaders load the exact same image:  True
The image loaded with decode_jpeg is equal to the others:  False
```

**Comment by pEtienn** [comment]
I forgot to add
> from tensorflow import io

in the code snippet.

**Comment by pEtienn** [comment]
There was another error in the test logic. I fixed it and put everything on this notebook: https://colab.research.google.com/drive/1__o37fZrczum7MVgAYl_nblXXZEORSDU?usp=sharing

**Comment by pEtienn** [comment]
I tested both TF 2.19 and 2.20

**Comment by Venkat6871** [comment]
I was able to reproduce the same issue with TensorFlow version 2.20.0 on colab. Here I am adding a [gist](https://colab.sandbox.google.com/gist/Venkat6871/042bf820a18b25793b5aaf35409e2402/102996_tf_2-20-0-v.ipynb) for reference.
Thank you!

**Comment by nileshau-afk** [comment]
Hi @pEtienn, 

Thank you for your patience, and apologies for the late reply. The observed differences in pixel values between images loaded by `tf.image.decode_jpeg` and other libraries like PIL or Matplotlib are expected due to variations in their underlying JPEG decoding implementations. Different libraries (e.g., TensorFlow often uses libjpeg-turbo, while PIL has its own optimized decoder or uses a different libjpeg version) can have subtle differences in their color space conversion matrices (e.g., YCbCr to RGB) and rounding algorithms. These numerical discrepancies are typically small and often perceptually insignificant.

If pixel-perfect consistency across different loading mechanisms is critical for your workflow, it is generally recommended to use a single, consistent image loading library throughout your pipeline. If you must use `tf.image.decode_jpeg`, be aware that minor numerical differences compared to other decoders are a known characteristic.

For more details on `tf.image.decode_jpeg`, you can refer to its official documentation: https://www.tensorflow.org/api_docs/python/tf/image/decode_jpeg
